File: AIReader/AIReader/Services/AI/ToolRunner.swift
import ComposableArchitecture
import Foundation

/// Carries a conversation with the model, answering the tools it calls —
/// the dictionary, the book search and the web — until it answers in words.
enum ToolRunner {
    enum RunError: LocalizedError {
        case gaveUp

        var errorDescription: String? { "The model kept searching without answering." }
    }

    /// What the model may reach for. A missing corpus or dictionary simply
    /// answers that tool with nothing. The web is offered only when it is set
    /// up, since a search there may be paid for — and always under the mock,
    /// which answers it itself.
    struct Tools: Sendable {
        var scope: ReadingScope = .none
        var dictionary: DictionaryClient?
        var web = WebSearchSettings()
    }

    /// How many rounds of tool calls the model gets before it must answer.
    static let budget = 3

    static func converse(
        settings: AISettings,
        messages: inout [ChatMessage],
        tools: [[String: Any]],
        jsonMode: Bool,
        available: Tools
    ) async throws -> ChatMessage {
        let offered = offered(tools, settings: settings, available: available)
        for _ in 0...budget {
            let reply = try await ChatAPI.chat(
                settings: settings,
                messages: messages,
                tools: offered,
                jsonMode: jsonMode
            )
            guard let calls = reply.toolCalls, !calls.isEmpty else { return reply }
            messages.append(reply)
            for call in calls {
                messages.append(.toolResult(await answer(call, settings: settings, available), callID: call.id))
            }
        }
        throw RunError.gaveUp
    }

    /// `tools` with the web added when it can be answered.
    private static func offered(_ tools: [[String: Any]], settings: AISettings, available: Tools) -> [[String: Any]] {
        guard settings.usesMock || available.web.isConfigured else { return tools }
        return tools + [WebSearchTool.tool]
    }

    private static func answer(_ call: ChatMessage.ToolCall, settings: AISettings, _ available: Tools) async -> String {
        switch call.function.name {
        case ExplanationPrompt.toolName:
            let word = ToolSchema.argument("word", in: call.function.arguments) ?? ""
            guard let dictionary = available.dictionary else {
                return DictionaryLookup(query: word).summary
            }
            return await dictionary.lookup(word).summary

        case SearchTool.toolName:
            let query = SearchTool.query(in: call.function.arguments)
            guard let corpus = available.scope.corpus, !query.isEmpty else {
                return SearchTool.summary(query: query, hits: [], severalBooks: false)
            }
            let hits = await corpus.search(query, limit: SearchTool.passageLimit, upTo: available.scope.upTo)
            return SearchTool.summary(query: query, hits: hits, severalBooks: corpus.severalBooks)

        case WebSearchTool.toolName:
            let query = WebSearchTool.query(in: call.function.arguments)
            guard !query.isEmpty else { return WebSearchTool.summary(query: query, output: nil) }
            // A failed search is told to the model, which can still answer
            // from what it has, rather than failing the whole conversation.
            do {
                // Pages in the book's language: a French name wants the French page.
                let language = available.scope.corpus?.books.first?.language ?? ""
                let output = settings.usesMock
                    ? MockAI.webSearch(query)
                    : try await WebSearch.search(settings: available.web, query: query, language: language)
                return WebSearchTool.summary(query: query, output: output)
            } catch {
                return "The web search failed: \(error.localizedDescription)"
            }

        default:
            return "There is no tool called “\(call.function.name)”."
        }
    }
}

File: AIReaderKindle/src/Services/ToolRunner.cpp
#include "ToolRunner.hpp"

#include "../Domain/AI/DictionaryTool.hpp"
#include "../Domain/AI/MockAI.hpp"
#include "../Domain/AI/SearchTool.hpp"
#include "../Domain/AI/WebSearchTool.hpp"
#include "ChatApi.hpp"
#include "DictionaryDatabase.hpp"
#include "WebSearch.hpp"

namespace ToolRunner {

namespace {

/// `tools` with the web added when it can be answered.
Json offered(const Json& tools, const AiSettings& settings, const Tools& available) {
    if (!settings.usesMock() && !available.web.isConfigured()) return tools;
    Json all = Json::array();
    for (const auto& tool : tools.items()) all.push(tool);
    all.push(WebSearchTool::tool());
    return all;
}

std::string answer(const ChatMessage::ToolCall& call, const AiSettings& settings, const Tools& available) {
    if (call.name == DictionaryTool::toolName) {
        std::string word = DictionaryTool::word(call.arguments);
        return DictionaryDatabase::shared().lookup(word, available.packs).summary();
    }
    if (call.name == SearchTool::toolName) {
        std::string query = SearchTool::query(call.arguments);
        const ReadingScope& scope = available.scope;
        if (!scope.corpus || query.empty()) return SearchTool::summary(query, {}, false);
        auto hits = scope.corpus->search(query, SearchTool::passageLimit, scope.upTo);
        return SearchTool::summary(query, hits, scope.corpus->severalBooks());
    }
    if (call.name == WebSearchTool::toolName) {
        std::string query = WebSearchTool::query(call.arguments);
        if (query.empty()) return WebSearchTool::summary(query, Json());
        // A failed search is told to the model, which can still answer
        // from what it has, rather than failing the whole conversation.
        try {
            // Pages in the book's language: a French name wants the French page.
            const auto& books = available.scope.corpus ? available.scope.corpus->books() : std::vector<Book>{};
            std::string language = books.empty() ? "" : books.front().language;
            Json output = settings.usesMock() ? MockAI::webSearch(query) : WebSearch::search(available.web, query, language);
            return WebSearchTool::summary(query, output);
        } catch (const std::exception& failure) {
            return std::string("The web search failed: ") + failure.what();
        }
    }
    return "There is no tool called “" + call.name + "”.";
}

}  // namespace

ChatMessage converse(
    const AiSettings& settings,
    std::vector<ChatMessage>& messages,
    const Json& tools,
    bool jsonMode,
    const Tools& available)
{
    Json all = offered(tools, settings, available);
    for (int round = 0; round <= budget; ++round) {
        ChatMessage reply = ChatApi::chat(settings, messages, all, jsonMode);
        if (reply.toolCalls.empty()) return reply;
        messages.push_back(reply);
        for (const auto& call : reply.toolCalls) {
            messages.push_back(ChatMessage::toolResult(answer(call, settings, available), call.id));
        }
    }
    throw ChatApi::Error("The model kept searching without answering.");
}

}  // namespace ToolRunner
